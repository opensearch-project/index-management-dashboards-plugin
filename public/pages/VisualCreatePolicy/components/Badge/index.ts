/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import Badge from "./Badge";

export default Badge;
