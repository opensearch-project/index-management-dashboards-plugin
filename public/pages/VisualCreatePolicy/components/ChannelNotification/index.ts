/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import ChannelNotification from "./ChannelNotification";

export default ChannelNotification;
