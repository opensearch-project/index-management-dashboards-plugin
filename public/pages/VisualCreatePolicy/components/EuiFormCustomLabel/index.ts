/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import EuiFormCustomLabel from "./EuiFormCustomLabel";

export default EuiFormCustomLabel;
