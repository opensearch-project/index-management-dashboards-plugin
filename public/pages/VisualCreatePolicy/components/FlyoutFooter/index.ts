/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import FlyoutFooter from "./FlyoutFooter";

export default FlyoutFooter;
