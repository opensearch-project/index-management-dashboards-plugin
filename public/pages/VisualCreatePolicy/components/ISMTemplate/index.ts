/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import ISMTemplate from "./ISMTemplate";

export default ISMTemplate;
