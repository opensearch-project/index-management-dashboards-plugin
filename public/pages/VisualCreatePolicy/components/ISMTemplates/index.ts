/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import ISMTemplates from "./ISMTemplates";

export default ISMTemplates;
