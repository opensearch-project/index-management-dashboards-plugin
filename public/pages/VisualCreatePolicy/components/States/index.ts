/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import States from "./States";

export default States;
