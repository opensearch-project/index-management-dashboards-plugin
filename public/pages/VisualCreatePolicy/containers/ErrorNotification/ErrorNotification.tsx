/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React, { ChangeEvent, Component } from "react";
import { EuiLink, EuiIcon, EuiFlexGroup, EuiFlexItem, EuiText } from "@elastic/eui";
import { ContentPanel } from "../../../../components/ContentPanel";
import "brace/theme/github";
import "brace/mode/json";
import { FeatureChannelList } from "../../../../../server/models/interfaces";
import { NotificationService } from "../../../../services";
import { ErrorNotification as IErrorNotification } from "../../../../../models/interfaces";
import { getErrorMessage } from "../../../../utils/helpers";
import { CoreServicesContext } from "../../../../components/core_services";
import ChannelNotification from "../../components/ChannelNotification";
import LegacyNotification from "../../components/LegacyNotification";
import { ERROR_NOTIFICATION_DOCUMENTATION_URL } from "../../../../utils/constants";

export interface ErrorNotificationProps {
  errorNotification: IErrorNotification | undefined;
  errorNotificationJsonString: string;
  onChangeChannelId: (value: ChangeEvent<HTMLSelectElement>) => void;
  onChangeMessage: (value: ChangeEvent<HTMLTextAreaElement>) => void;
  onChangeErrorNotificationJsonString: (str: string) => void;
  onSwitchToChannels: () => void;
  notificationService: NotificationService;
}

interface ErrorNotificationState {
  channels: FeatureChannelList[];
  loadingChannels: boolean;
}

export default class ErrorNotification extends Component<ErrorNotificationProps, ErrorNotificationState> {
  static contextType = CoreServicesContext;
  constructor(props: ErrorNotificationProps) {
    super(props);

    this.state = {
      channels: [],
      loadingChannels: true,
    };
  }

  componentDidMount = async (): Promise<void> => {
    await this.getChannels();
  };

  getChannels = async (): Promise<void> => {
    this.setState({ loadingChannels: true });
    try {
      const { notificationService } = this.props;
      const response = await notificationService.getChannels();
      if (response.ok) {
        this.setState({ channels: response.response.channel_list });
      } else {
        this.context.notifications.toasts.addDanger(`Could not load notification channels: ${response.error}`);
      }
    } catch (err) {
      this.context.notifications.toasts.addDanger(getErrorMessage(err, "Could not load the notification channels"));
    }
    this.setState({ loadingChannels: false });
  };

  render() {
    const {
      errorNotification,
      errorNotificationJsonString,
      onChangeChannelId,
      onChangeMessage,
      onChangeErrorNotificationJsonString,
      onSwitchToChannels,
    } = this.props;
    const { channels, loadingChannels } = this.state;
    const hasDestination = !!errorNotification?.destination;

    let content = (
      <ChannelNotification
        channelId={errorNotification?.channel?.id || ""}
        channels={channels}
        loadingChannels={loadingChannels}
        message={errorNotification?.message_template?.source || ""}
        onChangeChannelId={onChangeChannelId}
        onChangeMessage={onChangeMessage}
        getChannels={this.getChannels}
      />
    );

    // If we have a destination in the error notification then it's either an older policy or they created through the API
    if (hasDestination) {
      content = (
        <LegacyNotification
          notificationJsonString={errorNotificationJsonString}
          onChangeNotificationJsonString={onChangeErrorNotificationJsonString}
          onSwitchToChannels={onSwitchToChannels}
        />
      );
    }

    return (
      <ContentPanel
        bodyStyles={{ padding: "initial" }}
        title={
          <EuiFlexGroup gutterSize="xs" alignItems="center">
            <EuiFlexItem grow={false}>
              <EuiText>
                <h3>Error notification</h3>
              </EuiText>
            </EuiFlexItem>
            <EuiFlexItem>
              <EuiText color="subdued">
                <i> â€“ optional</i>
              </EuiText>
            </EuiFlexItem>
          </EuiFlexGroup>
        }
        titleSize="s"
        subTitleText={
          <EuiText color="subdued" size="s" style={{ padding: "5px 0px" }}>
            <p style={{ fontWeight: 200 }}>
              You can set up an error notification for when a policy execution fails.{" "}
              <EuiLink href={ERROR_NOTIFICATION_DOCUMENTATION_URL} target="_blank" rel="noopener noreferrer">
                Learn more
              </EuiLink>
            </p>
          </EuiText>
        }
      >
        <div style={{ padding: "10px 0px 0px 10px" }}>{content}</div>
      </ContentPanel>
    );
  }
}
