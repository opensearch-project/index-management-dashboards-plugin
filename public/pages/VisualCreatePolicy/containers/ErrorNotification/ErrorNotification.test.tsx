/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from "react";
import "@testing-library/jest-dom/extend-expect";
import { render } from "@testing-library/react";
import ErrorNotification, { ErrorNotificationProps } from "./ErrorNotification";
import { ServicesConsumer, ServicesContext } from "../../../../services";
import { browserServicesMock, coreServicesMock } from "../../../../../test/mocks";
import { BrowserServices } from "../../../../models/interfaces";
import { ErrorNotification as IErrorNotification } from "../../../../../models/interfaces";
import { CoreServicesContext } from "../../../../components/core_services";

function ErrorNotificationWrapper(props: ErrorNotificationProps) {
  return (
    <CoreServicesContext.Provider value={coreServicesMock}>
      <ErrorNotification {...props} />
    </CoreServicesContext.Provider>
  );
}

function renderErrorNotification(errorNotification: IErrorNotification) {
  return {
    ...render(
      <ServicesContext.Provider value={browserServicesMock}>
        <ServicesConsumer>
          {(services: BrowserServices | null) =>
            services && (
              <ErrorNotificationWrapper
                errorNotification={errorNotification}
                errorNotificationJsonString={""}
                onChangeChannelId={() => {}}
                onChangeMessage={() => {}}
                onChangeErrorNotificationJsonString={() => {}}
                onSwitchToChannels={() => {}}
                notificationService={services?.notificationService}
              />
            )
          }
        </ServicesConsumer>
      </ServicesContext.Provider>
    ),
  };
}

describe("<ErrorNotification /> spec", () => {
  it("renders the component", () => {
    const { container } = render(
      <ErrorNotificationWrapper
        errorNotification={{ channel: { id: "some_id" }, message_template: { source: "some source message" } }}
        errorNotificationJsonString={""}
        onChangeChannelId={() => {}}
        onChangeMessage={() => {}}
        onChangeErrorNotificationJsonString={() => {}}
        onSwitchToChannels={() => {}}
        notificationService={browserServicesMock.notificationService}
      />
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it("renders the channel ui editor for channels", () => {
    const errorNotification = { channel: { id: "some_id" }, message_template: { source: "some source message" } };
    const { queryByTestId } = renderErrorNotification(errorNotification);

    expect(queryByTestId("channel-notification-refresh")).not.toBeNull();
    expect(queryByTestId("create-policy-legacy-notification")).toBeNull();
  });

  it("renders the json legacy editor for destinations", () => {
    const errorNotification = { destination: { slack: { url: "https://slack.com" } }, message_template: { source: "some source message" } };
    const { queryByTestId } = renderErrorNotification(errorNotification);

    expect(queryByTestId("channel-notification-refresh")).toBeNull();
    expect(queryByTestId("create-policy-legacy-notification")).not.toBeNull();
  });
});
