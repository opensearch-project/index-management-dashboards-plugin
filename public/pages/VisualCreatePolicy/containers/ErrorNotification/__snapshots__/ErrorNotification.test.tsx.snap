// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ErrorNotification /> spec renders the component 1`] = `
<div
  class="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow"
  style="padding-left: 0px; padding-right: 0px;"
>
  <div
    class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsFlexStart euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
    style="padding: 0px 16px;"
  >
    <div
      class="euiFlexItem"
    >
      <div
        class="euiFlexGroup euiFlexGroup--gutterExtraSmall euiFlexGroup--alignItemsCenter euiFlexGroup--directionRow euiFlexGroup--responsive"
      >
        <div
          class="euiFlexItem euiFlexItem--flexGrowZero"
        >
          <div
            class="euiText euiText--medium"
          >
            <h3>
              Error notification
            </h3>
          </div>
        </div>
        <div
          class="euiFlexItem"
        >
          <div
            class="euiText euiText--medium"
          >
            <div
              class="euiTextColor euiTextColor--subdued"
            >
              <i>
                 – optional
              </i>
            </div>
          </div>
        </div>
      </div>
      <div
        class="euiText euiText--small"
        style="padding: 5px 0px;"
      >
        <div
          class="euiTextColor euiTextColor--subdued"
        >
          <p
            style="font-weight: 200;"
          >
            You can set up an error notification for when a policy execution fails.
             
            <a
              class="euiLink euiLink--primary"
              href="https://opensearch.org/docs/im-plugin/ism/policies/#error-notifications"
              rel="noopener noreferrer"
              target="_blank"
            >
              Learn more
              EuiIconMock
              <span
                class="euiScreenReaderOnly"
              >
                (opens in a new tab or window)
              </span>
            </a>
          </p>
        </div>
      </div>
    </div>
  </div>
  <hr
    class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginXSmall"
  />
  <div
    style="padding: initial;"
  >
    <div
      style="padding: 10px 0px 0px 10px;"
    >
      <div
        class="euiText euiText--medium"
        style="margin-bottom: 5px;"
      >
        <h5
          class="euiFormLabel "
          style="margin-bottom: 2px;"
        >
          Channel ID
        </h5>
        <p>
           
        </p>
      </div>
      <div
        class="euiFlexGroup euiFlexGroup--gutterSmall euiFlexGroup--directionRow euiFlexGroup--responsive"
        style="max-width: 600px;"
      >
        <div
          class="euiFlexItem"
        >
          <div
            class="euiFormRow euiFormRow--compressed"
            id="some_html_id-row"
          >
            <div
              class="euiFormRow__fieldWrapper"
            >
              <div
                class="euiFormControlLayout euiFormControlLayout--compressed"
              >
                <div
                  class="euiFormControlLayout__childrenWrapper"
                >
                  <select
                    class="euiSelect euiSelect--compressed euiSelect-isLoading"
                    data-test-subj="create-policy-notification-channel-id"
                    id="some_html_id"
                    placeholder="Select channel ID"
                  >
                    <option
                      disabled=""
                      hidden=""
                      style="display: none;"
                      value=""
                    >
                       
                    </option>
                  </select>
                  <div
                    class="euiFormControlLayoutIcons euiFormControlLayoutIcons--right"
                  >
                    <span
                      class="euiLoadingSpinner euiLoadingSpinner--medium"
                    />
                    <span
                      class="euiFormControlLayoutCustomIcon"
                    >
                      EuiIconMock
                    </span>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="euiFlexItem euiFlexItem--flexGrowZero"
        >
          <button
            class="euiButton euiButton--primary euiButton--small euiButton-isDisabled refresh-button"
            data-test-subj="channel-notification-refresh"
            disabled=""
            type="button"
          >
            <span
              class="euiButtonContent euiButton__content"
            >
              EuiIconMock
              <span
                class="euiButton__text"
              />
            </span>
          </button>
        </div>
        <div
          class="euiFlexItem euiFlexItem--flexGrowZero"
        >
          <a
            class="euiButton euiButton--primary euiButton--small"
            href="notifications-dashboards#/channels"
            rel="noopener noreferrer"
            target="_blank"
          >
            <span
              class="euiButtonContent euiButton__content"
            >
              EuiIconMock
              <span
                class="euiButton__text"
              >
                Manage channels
              </span>
            </span>
          </a>
        </div>
      </div>
      <div
        class="euiSpacer euiSpacer--m"
      />
      <div
        class="euiText euiText--medium"
        style="margin-bottom: 5px;"
      >
        <h5
          class="euiFormLabel "
          style="margin-bottom: 2px;"
        >
          Notification message
        </h5>
        <p>
           
          <span
            style="font-weight: 200; font-size: 12px;"
          >
            Embed variables in your message using Mustache template.
             
          </span>
        </p>
      </div>
      <div
        class="euiFormRow euiFormRow--compressed"
        id="some_html_id-row"
      >
        <div
          class="euiFormRow__fieldWrapper"
        >
          <textarea
            class="euiTextArea euiTextArea--resizeVertical euiTextArea--compressed"
            data-test-subj="create-policy-notification-message"
            id="some_html_id"
            placeholder="The index {{ctx.index}} failed during policy execution."
            rows="3"
            style="min-height: 150px;"
          >
            some source message
          </textarea>
        </div>
      </div>
    </div>
  </div>
</div>
`;
